Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1669 @@
+            }
+
+            _guid = Guid.Parse(referenceStringParts[0].Substring(3));
+            var versionSplit = referenceStringParts[1].Split(new char[] { '.' }, 2);
+            if (versionSplit.Length != 2)
+            {
+                throw new ArgumentException($"Invalid reference string got {referenceStringRaw}.  Invalid version string.");
+            }
+            _majorVersion = uint.Parse(versionSplit[0], NumberStyles.AllowHexSpecifier);
+            _minorVersion = uint.Parse(versionSplit[1], NumberStyles.AllowHexSpecifier);
+
+            _lcid = uint.Parse(referenceStringParts[2], NumberStyles.AllowHexSpecifier);
+            _path = referenceStringParts[3];
+            _name = referenceStringParts[4];
+        }
+
+        public void Document(StringLineBuilder output)
+        {
+            output.AppendLine("- VBE Reference: " + Name + " [path: " + Path + ", majorVersion: " + MajorVersion +
+                                ", minorVersion: " + MinorVersion + ", guid: " + GUID + ", lcid: " + LCID + "]");
+        }
+    }
+
+    /// <summary>
+    /// A class that represents a variable or field within a typeinfo
+    /// </summary>
+    public class TypeInfoVar : IDisposable
+    { 
+        private TypeInfoWrapper _typeInfo;
+        private ComTypes.VARDESC _varDesc;
+        private IntPtr _varDescPtr;
+        private string _name;
+
+        public string Name { get => _name; }
+
+        public TypeInfoVar(TypeInfoWrapper typeInfo, int index)
+        {
+            _typeInfo = typeInfo;
+
+            ((ComTypes.ITypeInfo)_typeInfo).GetVarDesc(index, out _varDescPtr);
+            _varDesc = StructHelper.ReadStructure<ComTypes.VARDESC>(_varDescPtr);
+            
+            int _cNames = 0;
+            var _names = new string[1];
+            if (_varDesc.memid != (int)TypeLibConsts.MEMBERID_NIL)
+            {
+                _cNames = 0;
+                ((ComTypes.ITypeInfo)_typeInfo).GetNames(_varDesc.memid, _names, _names.Length, out _cNames);
+                _name = _names[0];
+            }
+            else
+            {
+                _name = "{unknown}";     // VBA Constants appear in the typelib with no name
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_varDescPtr != IntPtr.Zero) ((ComTypes.ITypeInfo)_typeInfo).ReleaseVarDesc(_varDescPtr);
+            _varDescPtr = IntPtr.Zero;
+        }
+
+        public void Document(StringLineBuilder output) 
+        {
+            output.AppendLine("- field: " + _name + " [id 0x" + _varDesc.memid.ToString("X") + "]");
+        }
+    }
+    
+    /// <summary>
+    /// A base class for exposing an enumerable collection through an index based accessor
+    /// </summary>
+    /// <typeparam name="TItem">the collection element type</typeparam>
+    public class IIndexedCollection<TItem> : IEnumerable<TItem>
```

Comment:
This looks like an `abstract class` to me. I'd add the `abstract` to the class, replace `virtual` with `abstract` for the `Count` and `GetItemByIndex`. The difference is that all derived classes now *must* provide their implementation of the `abstract` members as opposed to *may* override an existing implementation. Furthermore, you can remove the `throw new exception` body; abstract methods only need the signatures defined.

Reply:
the convention in the codebase is then to call the class `IndexedCollectionBase`, as well ...

Reply:
thanks... that's the effect I wanted (and would normally do equivalent in C++), but wasn't sure how to do that in C#, so thanks.  did I mention I don't know c#? ... :+1: